File: AllegroDotNet.Wrapper/src/AlSystem.cpp
#include "PCH.h"

#include "Al.h"
#include "AllegroConfig.h"
#include "AllegroPath.h"
#include "AlNative.h"
#include "SystemID.h"

#include <msclr/marshal.h>

namespace AllegroDotNet::Wrapper
{
	using namespace msclr::interop;

	AllegroConfig^ Al::GetSystemConfig()
	{
		auto* nativeSystemConfig = al_get_system_config();
		return gcnew AllegroConfig(nativeSystemConfig);
	}

	AllegroPath^ Al::GetStandardPath(StandardPath id)
	{
		auto nativeId = (int)id;
		auto* nativePath = al_get_standard_path(nativeId);
		return gcnew AllegroPath(nativePath);
	}

	Boolean Al::Init()
	{
		return al_init();
	}

	Boolean Al::IsSystemInstalled()
	{
		return al_is_system_installed();
	}

	Int32 Al::GetCpuCount()
	{
		return al_get_cpu_count();
	}

	Int32 Al::GetRamSize()
	{
		return al_get_ram_size();
	}

	String^ Al::GetAppName()
	{
		auto* nativeAppName = al_get_app_name();
		return gcnew String(nativeAppName);
	}

	String^ Al::GetOrgName()
	{
		auto* nativeOrgName = al_get_org_name();
		return gcnew String(nativeOrgName);
	}

	SystemID Al::GetSystemID()
	{
		auto nativeSystemID = al_get_system_id();
		return (SystemID)nativeSystemID;
	}

	UInt32 Al::GetAllegroVersion()
	{
		return al_get_allegro_version();
	}

	void Al::RegisterAssertHandler(Action<String^, String^, Int32, String^>^ handler)
	{
		ManagedAssertHandler = handler;
		if (handler == nullptr)
		{
			al_register_assert_handler(nullptr);
		}
		else
		{
			al_register_assert_handler(AlNative::native_assert_handler);
		}
	}

	void Al::RegisterTraceHandler(Action<String^>^ handler)
	{
		ManagedTraceHandler = handler;
		if (handler == nullptr)
		{
			al_register_trace_handler(nullptr);
		}
		else
		{
			al_register_trace_handler(AlNative::native_trace_handler);
		}
	}

	void Al::SetAppName(String^ name)
	{
		marshal_context marshalContext;
		const auto* nativeAppName = marshalContext.marshal_as<const char*>(name);
		al_set_app_name(nativeAppName);
	}

	void Al::SetExeName(String^ name)
	{
		marshal_context marshalContext;
		const auto* nativeExeName = marshalContext.marshal_as<const char*>(name);
		al_set_app_name(nativeExeName);
	}

	void Al::SetOrgName(String^ name)
	{
		marshal_context marshalContext;
		const auto* nativeOrgName = marshalContext.marshal_as<const char*>(name);
		al_set_org_name(nativeOrgName);
	}

	void Al::UninstallSystem()
	{
		al_uninstall_system();
	}
}

File: Source/AllegroDotNet/Al.System.cs
using SubC.AllegroDotNet.Enums;
using SubC.AllegroDotNet.Models;
using SubC.AllegroDotNet.Native;

namespace SubC.AllegroDotNet;

/// <summary>
/// This static class contains the Allegro 5 library methods.
/// </summary>
public static partial class Al
{
  /// <summary>
  /// Attempts to initialize the Allegro system. The versions tried are limited to versions in the
  /// <see cref="LibraryVersion"/> enumeration.
  /// </summary>
  /// <returns>True if Allegro was initialized, otherwise false.</returns>
  public static bool Init()
  {
    foreach (LibraryVersion libraryVersion in Enum.GetValues(typeof(LibraryVersion)))
    {
      if (InstallSystem(libraryVersion))
        return true;
    }

    return false;
  }

  /// <summary>
  /// Initialize the Allegro system. No other Allegro functions can be called before this (with one or two exceptions).
  /// </summary>
  /// <param name="version">The library version of Allegro to initialize.</param>
  /// <returns>True on success, otherwise false.</returns>
  public static bool InstallSystem(LibraryVersion version)
  {
    return InstallSystem((int)version);
  }

  /// <summary>
  /// Initialize the Allegro system. No other Allegro functions can be called before this (with one or two exceptions).
  /// </summary>
  /// <param name="version">The library version of Allegro to initialize, packed into a integer.</param>
  /// <returns>True on success, otherwise false.</returns>
  public static bool InstallSystem(int version)
  {
    return Interop.Core.AlInstallSystem(version, null) != 0;
  }

  /// <summary>
  /// Closes down the Allegro system.
  /// </summary>
  public static void UninstallSystem()
  {
    Interop.Core.AlUninstallSystem();
  }

  /// <summary>
  /// Returns true if Allegro is initialized, otherwise returns false.
  /// </summary>
  /// <returns>True if Allegro is initialized, otherwise false.</returns>
  public static bool IsSystemInstalled()
  {
    return Interop.Core.AlIsSystemInstalled() != 0;
  }

  /// <summary>
  /// Returns the (compiled) version of the Allegro library, packed into a single integer.
  /// This value can be cast to (or from) a value of <see cref="LibraryVersion"/>, if it defined
  /// in that enumeration; they use the same integer values.
  /// </summary>
  /// <returns>The compiled version of the Allegro library, packed into a single integer.</returns>
  public static uint GetAllegroVersion()
  {
    return Interop.Core.AlGetAllegroVersion();
  }

  public static AllegroPath? GetStandardPath(StandardPath path)
  {
    var pointer = Interop.Core.AlGetStandardPath((int)path);
    return NativePointer.Create<AllegroPath>(pointer);
  }

  public static void SetExeName(string name)
  {
    using var nativeName = new CStringAnsi(name);
    Interop.Core.AlSetExeName(nativeName.Pointer);
  }

  public static void SetAppName(string name)
  {
    using var nativeName = new CStringAnsi(name);
    Interop.Core.AlSetAppName(nativeName.Pointer);
  }

  public static void SetOrgName(string name)
  {
    using var nativeName = new CStringAnsi(name);
    Interop.Core.AlSetOrgName(nativeName.Pointer);
  }

  public static string? GetAppName()
  {
    var pointer = Interop.Core.AlGetAppName();
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static string? GetOrgName()
  {
    var pointer = Interop.Core.AlGetOrgName();
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static AllegroConfig? GetSystemConfig()
  {
    var pointer = Interop.Core.AlGetSystemConfig();
    return NativePointer.Create<AllegroConfig>(pointer);
  }

  public static SystemID GetSystemID()
  {
    return (SystemID)Interop.Core.AlGetSystemId();
  }

  public static void RegisterAssertHandler(Delegates.RegisterAssertHandlerDelegate? handler)
  {
    Interop.Core.AlRegisterAssertHandler(handler);
  }

  public static void RegisterTraceHandler(Delegates.RegisterTraceHandlerDelegate? handler)
  {
    Interop.Core.AlRegisterTraceHandler(handler);
  }

  public static int GetCpuCount()
  {
    return Interop.Core.AlGetCpuCount();
  }

  public static int GetRamSize()
  {
    return Interop.Core.AlGetRamSize();
  }
}
